proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+
+### Plugin API
+
+The new `PluginContext` structure in the `PackagePlugin` API is defined as:
+
+```swift
+/// Provides information about the package for which the plugin is invoked,
+/// as well as contextual information based on the plugin's stated intent
+/// and requirements.
+public struct PluginContext {
+    /// Information about the package to which the plugin is being applied.
+    public let package: Package
+
+    /// The path of a writable directory into which the plugin or the build
+    /// commands it constructs can write anything it wants. This could include
+    /// any generated source files that should be processed further, and it
+    /// could include any caches used by the build tool or the plugin itself.
+    /// The plugin is in complete control of what is written under this di-
+    /// rectory, and the contents are preserved between builds.
+    ///
+    /// A plugin would usually create a separate subdirectory of this directory
+    /// for each command it creates, and the command would be configured to
+    /// write its outputs to that directory. The plugin may also create other
+    /// directories for cache files and other file system content that either
+    /// it or the command will need.
+    public let pluginWorkDirectory: Path
+
+    /// The path of the directory into which built products associated with
+    /// the target are written.
+    public let builtProductsDirectory: Path
+
+    /// Looks up and returns the path of a named command line executable tool.
+    /// The executable must be provided by an executable target or a binary
+    /// target on which the package plugin target depends. This function throws
+    /// an error if the tool cannot be found. The lookup is case sensitive.
+    public func tool(named name: String) throws -> Tool
+    
+    /// Information about a particular tool that is available to a plugin.
+    public struct Tool {
+        /// Name of the tool (suitable for display purposes).
+        public let name: String
+
+        /// Full path of the built or provided tool in the file system.
+        public let path: Path
+    }
+}
+```
+
+The `package` property is a reference to the top-level package to which the plugin is being applied. Through it, the entire subgraph of resolved packages reachable from it can be accessed.
+
+The properties related to looking up tools with a particular name are unchanged from the original SE-0303 proposal.
+
+This has the effect of factoring out all information related to the package and target, and it puts them into its own directed acyclic graph consisting of the following types:
+
+```swift
+/// Represents a single package in the graph (either the root or a dependency).
+public class Package {
+    /// The name of the package (for display purposes only).
+    public let name: String
```

Comment:
should this be called `displayName` to make the intent clearer?

Reply:
Yes, that’s a good point.  I can definitely see people using this for purposes that it wasn’t intended for.

I wonder if we should do the same with the target — I already split out `moduleName` from the target’s `name` (especially since we should one day allow the manifest to customize it), but separating out product name from display name there too might be useful.